janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java

```diff
@@ -6871,4 +6871,27 @@ public void performReindexAndVerifyEdgeCount(String indexName, String edgeLabel,
                 g.V(v).properties("vtName").next().value(), indexName, dirs[i], propValues[i]));
         }
     }
+
+    @Test
+    public void testMultipleOrClauses() {
```

Comment:
Nice, I just verified this test fails on master.

I haven't got a chance to look at the implementation yet, but I think there are a few other scenarios worth testing (you probably have already tested, but I would appreciate it if you could add them into tests):
1) Nested OR clauses, or even deeply nested OR clauses.
2) Multiple OR clauses that hit one or more composite indexes.
3) Multiple OR clauses that hit one or more mixed indexes (JanusGraphIndexTest).

Reply:
Added some more test cases. Couldn't actually put together a test case that hits a composite index. It seems like when there are multiple or steps, it won't hit any composite index, but I haven't dug deep into this yet. Is it possible for CNF queries to use composite indexes?

Reply:
Whether to use composite index for multiple OR steps is a tricky problem. For example, for a query 
```java
tx.traversal().V()
    .or(__.has("a", true), __.has("b", true))
    .or(__.has("c", false), __.has("d", true))
```
If there are composite indexes for "a", "b", "c", and "d", respectively, should we make use of composite indexes or not? We could use composite indexes for each of them and do intersections, or we could only use index for "a", "b" and do in-memory filtering for "c" and "d".... Without histograms it's hard to say. In this case, I still believe either way is better than a full scan.

Currently, a single OR query
```java
tx.traversal().V()
    .or(__.has("a", true), __.has("b", true))
```
will try to make use of indexes as much as possible. If both "a" and "b" have composite indexes (and no mixed index), then composite indexes will be used and then results will be merged. Thus, if now nested OR query does not use composite indexes, something might be wrong (or the design itself does not consider this). I haven't checked your code yet thus not sure.

Reply:
The design itself doesn't consider this, since before this PR nested OR queries were squashed into a single OR statement so it did not have to handle such cases for composite indexes and my changes do not make any significant changes to the underlying logic.